src/main/java/it/smartcommunitylab/aac/tos/TosOnAccessFilter.java

```diff
@@ -0,0 +1,160 @@
+	static final String[] SKIP_URLS = { "/api/**", "/html/**", "/js/**", "/lib/**", "/fonts/**", "/italia/**",
+			"/i18n/**" };
+	private final Logger logger = LoggerFactory.getLogger(getClass());
+
+	private final RequestCache requestCache;
+
+	private RedirectStrategy redirectStrategy = new DefaultRedirectStrategy();
+	private RequestMatcher requestMatcher;
+
+	private final RealmManager realmManager;
+	private final MyUserManager userManager;
+
+	public TosOnAccessFilter(RealmManager realmManager, MyUserManager userManager) {
+		// init request cache as store
+		HttpSessionRequestCache cache = new HttpSessionRequestCache();
+		this.requestCache = cache;
+		this.realmManager = realmManager;
+		this.userManager = userManager;
+		this.requestMatcher = buildRequestMatcher();
+	}
+
+	private RequestMatcher buildRequestMatcher() {
+		List<RequestMatcher> antMatchers = Arrays.stream(SKIP_URLS).map(u -> new AntPathRequestMatcher(u))
+				.collect(Collectors.toList());
+
+		return new NegatedRequestMatcher(new OrRequestMatcher(antMatchers));
+
+	}
+
+	public void setRedirectStrategy(RedirectStrategy redirectStrategy) {
+		this.redirectStrategy = redirectStrategy;
+	}
+
+	public void setRequestMatcher(RequestMatcher requestMatcher) {
+		this.requestMatcher = requestMatcher;
+	}
+
+	@Override
+	protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain chain)
+			throws ServletException, IOException {
+		if (requestMatcher.matches(request) && requiresProcessing(request)
+				&& !termsManagedRequestMatcher.matches(request)) {
+			logger.trace("process request for {}", request.getRequestURI());
+			UserAuthentication userAuth = (UserAuthentication) SecurityContextHolder.getContext().getAuthentication();
+			ExtendedAuthenticationToken token = CollectionUtils.firstElement(userAuth.getAuthentications());
+
+			if (token == null) {
+				logger.error("empty token on authentication success");
+				return;
+			}
+
+			String realm = token.getRealm();
+
+			try {
+				User user = userManager.getMyUser(realm);
+				if (!realm.equalsIgnoreCase(SystemKeys.REALM_GLOBAL) && !realm.equalsIgnoreCase(SystemKeys.REALM_SYSTEM)
+						&& !user.isTosAccepted()) {
+					Realm realmEntity = realmManager.findRealm(realm);
+
+					if ((realmEntity.getTosConfiguration().getConfiguration().containsKey("enableTOS"))
+							&& (boolean) realmEntity.getTosConfiguration().getConfiguration().get("enableTOS")
+							&& request.getSession().getAttribute("termsManaged") == null) {
+						String targetUrl = "/terms/"
+								+ realmEntity.getTosConfiguration().getConfiguration().get("approveTOS");
+						this.requestCache.saveRequest(request, response);
+						this.logger.debug("Redirect to {}", targetUrl);
+						this.redirectStrategy.sendRedirect(request, response, targetUrl);
+						return;
+					} else if (request.getSession().getAttribute("termsManaged") != null
```

Comment:
why you need the session to carry the form results? they should be sent to either a filter or a controller to parse, handle and update the user model. The only info needed here is that the user made *a choice* to avoid sending him back to the form.